File: Desktop/src/ca/mcgill/ecse321/urlms/controller/ManagerController.java
package ca.mcgill.ecse321.urlms.controller;

import java.sql.Date;
import java.util.Calendar;
import java.util.List;
import ca.mcgill.ecse321.urlms.application.URLMSApplication;
import ca.mcgill.ecse321.urlms.model.Assistant;
import ca.mcgill.ecse321.urlms.model.Associate;
import ca.mcgill.ecse321.urlms.model.Director;
import ca.mcgill.ecse321.urlms.model.Equipment;
import ca.mcgill.ecse321.urlms.model.Expense;
import ca.mcgill.ecse321.urlms.model.Expense.Type;
import ca.mcgill.ecse321.urlms.model.Inventory;
import ca.mcgill.ecse321.urlms.model.Lab;
import ca.mcgill.ecse321.urlms.model.ProgressUpdate;
import ca.mcgill.ecse321.urlms.model.Resource;
import ca.mcgill.ecse321.urlms.model.Staff;
import ca.mcgill.ecse321.urlms.model.Supply;
import ca.mcgill.ecse321.urlms.model.URLMS;
import ca.mcgill.ecse321.urlms.persistence.PersistenceXStream;

public class ManagerController {
	private URLMS urlms;
	private Lab singleLab;

	public ManagerController() {
		PersistenceXStream.initializeModelManager(PersistenceXStream.getFilename());
		// try {
		// loadLab();
		// } catch (InvalidInputException e) {
		// // TODO Auto-generated catch block
		// e.printStackTrace();
		// }
	}

	/**
	 * Method to add a User to the system
	 * 
	 * @param firstName
	 * @param lastName
	 * @param role
	 * @throws InvalidInputException
	 */
	public void addStaff(String firstName, String lastName, String role) throws InvalidInputException {
		String error = "";
		if (error.length() > 0) {
			throw new InvalidInputException(error.trim());
		}

		try {
			if (role == "Director") {
				if (singleLab.hasDirector()) {
					throw new InvalidInputException("The lab already has a director");
				}
				else {
					Staff staff = new Staff(firstName, lastName, createID(), singleLab);
					createDirector(firstName, lastName, staff, singleLab);
				}
			}
			else if (role == "Assistant") {
				Staff staff = new Staff(firstName, lastName, createID(), singleLab);

				Assistant assistant = new Assistant();
				staff.setRole(assistant);
				singleLab.addStaff(staff);
			}
			else if (role == "Associate") {
				Staff staff = new Staff(firstName, lastName, createID(), singleLab);

				Associate associate = new Associate();
				staff.setRole(associate);
				singleLab.addStaff(staff);
			}

		} catch (RuntimeException e) {
			throw new InvalidInputException(error);
		}

	}

	private int createID() {
		int iD = (int) (Math.random() * 100000);
		while (iD == 0) {
			iD = (int) (Math.random() * 100000);
		}
		int size = URLMSApplication.getURLMS().getSingleLab().getStaff().size();
		for (int i = 0; i < size; i++) {
			if (URLMSApplication.getURLMS().getSingleLab().getStaff().get(i).getID() == iD) {
				iD = (int) (Math.random() * 100000);
				i = 0;
			}
		}

		return iD;

	}

	/**
	 * This method removes a staff from the lab
	 * 
	 * @param aStaff staff to remove
	 */
	public void removeStaff(Staff aStaff) throws InvalidInputException {
		String error = "";

		try {
			// singleLab.removeStaff(aStaff);
			if (aStaff.getRole() instanceof Director) {
				singleLab.setDirector(null);
			}
			aStaff.getRole().delete();
			aStaff.delete();
		} catch (RuntimeException e) {
			throw new InvalidInputException(error);
		}
	}

	/**
	 * This method changes the role of a staff member
	 * 
	 * @param aStaff
	 * @param role
	 * @throws InvalidInputException
	 */
	public void editStaffRole(Staff aStaff, String role) throws InvalidInputException {
		String error = "";

		if (role == "Director" && singleLab.hasDirector()) {
			error = "The lab already has a director";
		}
		if (error.length() > 0) {
			throw new InvalidInputException(error.trim());
		}
		if (error.length() == 0) {
			try {
				if (role == "Director") {
					aStaff.getRole().delete();
					aStaff.setRole(new Director());
				}
				else if (role == "Assistant") {
					aStaff.getRole().delete();
					aStaff.setRole(new Assistant());
				}
				else if (role == "Associate") {
					aStaff.getRole().delete();
					aStaff.setRole(new Associate());
				}
			} catch (RuntimeException e) {
				throw new InvalidInputException(error);
			}
		}

	}

	public void setPassword(Staff aStaff, String password) {
		String error = "";

		if (password.length() == 0) {
			error = "There needs to be a password";
		}
		if (error.length() == 0) {
			aStaff.setPassword(password);
		}
	}

	/**
	 * This methods adds an equipement to the inventory
	 * 
	 * @param quantity of the equipment to be added
	 * @param name of the added equipment
	 * @throws InvalidInputException
	 */
	public void addEquipement(int quantity, String name) throws InvalidInputException {
		String error = "";
		Inventory currentInventory = singleLab.getInventory();

		List<Resource> resources = currentInventory.getResources();

		for (Resource resource : resources) {
			if (resource.getName().equals(name)) {
				error = "An equipment with that name already exists.";
			}
		}

		if (error.length() > 0) {
			throw new InvalidInputException(error.trim());
		}

		if (error.isEmpty()) {
			try {
				Resource newEquipment = new Equipment(quantity, name, currentInventory);
				currentInventory.addResource(newEquipment);
			} catch (RuntimeException e) {
				throw new InvalidInputException(error);
			}
		}
	}

	public boolean isFundingAccountSet() {
		if (singleLab.getFundingAccount().getAllocatedAmount() > 0) {
			return true;
		}
		return false;
	}

	public void setFundingAccount(int amount) throws InvalidInputException {
		String error = "";
		if (amount <= 0) {
			error = "The amount needs to be positive";
		}
		if (singleLab.getFundingAccount().getAllocatedAmount() > 0) {
			error = "The funding account has already been set";
		}
		if (error.length() == 0) {
			singleLab.getFundingAccount().setAllocatedAmount(amount);
		}
		else {
			throw new InvalidInputException(error);
		}
	}

	public void addFundingAccount(int amount) throws InvalidInputException {
		String error = "";
		if (amount <= 0) {
			error = "The amount needs to be positive";
		}
		if (error.length() == 0) {
			singleLab.getFundingAccount()
					.setAllocatedAmount(singleLab.getFundingAccount().getAllocatedAmount() + amount);
		}
		else {
			throw new InvalidInputException(error);
		}
	}

	/**
	 * This method adds a supply to the inventory
	 * 
	 * @param quantity of the supply
	 * @param name of the supply
	 * @throws InvalidInputException
	 */
	public void addSupply(int quantity, String name) throws InvalidInputException {
		String error = "";
		Inventory currentInventory = singleLab.getInventory();

		List<Resource> resources = currentInventory.getResources();

		for (Resource resource : resources) {
			// System.out.println(resource.getName());
			// System.out.println("my resource " + name);
			if ((resource.getName()).equals(name)) {
				error = "A supply with that name already exists.";
			}
		}

		if (error.length() > 0) {
			throw new InvalidInputException(error.trim());
		}
		if (error.isEmpty()) {
			try {
				Resource newSupply = new Supply(quantity, name, currentInventory);
				currentInventory.addResource(newSupply);
			} catch (RuntimeException e) {
				throw new InvalidInputException(error);
			}
		}
	}

	/**
	 * This method edits the quantity of a given resource
	 * 
	 * @param resource
	 * @param newQty
	 * @throws InvalidInputException
	 */
	public void editResource(Resource resource, int newQty) throws InvalidInputException {
		String error = "";
		if (newQty < 1) {
			error = "The quantity must be more than 0";
		}
		if (error.length() > 0) {
			throw new InvalidInputException(error.trim());
		}
		else if (error.isEmpty()) {
			resource.setQuantity(newQty);
		}
	}

	/**
	 * This method will remove a resource from the system
	 * 
	 * @param aResource
	 * @throws InvalidInputException
	 */
	public void removeResource(Resource aResource) throws InvalidInputException {
		String error = "";

		try {
			singleLab.getInventory().removeResource(aResource);
			aResource.delete();
		} catch (RuntimeException e) {
			throw new InvalidInputException(error);
		}
	}

	/**
	 * This method will create a director and set the lab's director to this specific director
	 * 
	 * @param firstName
	 * @param lastName
	 * @param staff
	 * @param lab
	 * @throws InvalidInputException
	 */
	public void createDirector(String firstName, String lastName, Staff staff, Lab lab) throws InvalidInputException {
		String error = "";
		if (lab.hasDirector()) {
			error = "You already have a Director";
		}
		if (error.length() > 0) {
			throw new InvalidInputException(error.trim());
		}
		try {
			Director director = new Director();
			staff.setRole(director);
			lab.addStaff(staff);
			lab.setDirector(director);
		} catch (RuntimeException e) {
			throw new InvalidInputException(e.getMessage());
		}
	}

	/**
	 * This method creates and adds a progress update to a specific staff member. it also adds the
	 * progress update to the lab
	 * 
	 * @param content
	 * @param aStaff
	 * @throws InvalidInputException
	 */
	public void createProgressUpdate(String content, Staff aStaff) throws InvalidInputException {
		String error = "";
		if (content.length() == 0) {
			error = "The progress update is empty";
		}
		if (error.length() > 0) {
			throw new InvalidInputException(error.trim());
		}
		else {
			try {
				Date now = new Date(Calendar.getInstance().getTimeInMillis());
				ProgressUpdate progress = new ProgressUpdate(content, aStaff.getFirstName(), aStaff.getLastName(), now,
						singleLab);
				aStaff.addProgressUpdate(progress);
			} catch (RuntimeException e) {
				throw new InvalidInputException(e.getMessage());
			}
		}
	}

	public void addExpense(Staff aStaff, String description, float value, String type) throws InvalidInputException {
		String error = "";
		if (description.length() == 0) {
			error = "The description cannot be empty";
		}
		if (type.length() == 0) {
			error = "The expense needs to be categorized";
		}
		if (value < 0) {
			error = "The amount cannot be negative";
		}
		if (error.length() == 0) {
			try {
				Date now = new Date(Calendar.getInstance().getTimeInMillis());
				Expense e = new Expense(description, value, now, singleLab.getFundingAccount());
				if (type == "Travel") {
					e.setType(Type.TRAVEL);
				}
				else if (type == "Salary") {
					e.setType(Type.SALARY);
				}
				else if (type == "Resource") {
					e.setType(Type.RESOURCE);
				}
				else if (type == "Other") {
					e.setType(Type.OTHER);
				}

				aStaff.addExpense(e);
			} catch (RuntimeException e) {
				throw new InvalidInputException(e.getMessage());
			}
		}
	}

	/**
	 * This method creates a lab and adds it to the URLMS application
	 * 
	 * @throws InvalidInputException
	 */
	public void createLab(String password) throws InvalidInputException {
		URLMS urlms = URLMSApplication.getURLMS();
		Lab lab = new Lab(password, urlms);
		singleLab = lab;
		urlms.setSingleLab(lab);
	}

	/**
	 * Method saves the URLMS to the persistence
	 */
	public void saveLab() {
		URLMSApplication.save();
	}

	public Lab loadLab() throws InvalidInputException {
    String error = "";
    if (!URLMSApplication.getURLMS().hasSingleLab()) {
      error = "The current URLMS does not have labs";
      throw new InvalidInputException(error.trim());
    }
    try {
      singleLab = URLMSApplication.getURLMS().getSingleLab();
      return singleLab;
    } catch (RuntimeException e) {
      throw new InvalidInputException(e.getMessage());
    }
  }
}

File: URLMS_WebApp/controller/ManagerController.php
<?php

if(!isset($_SESSION)){ //Set Session only once
    session_start();
}

$my_dir = dirname(__FILE__);
require_once $my_dir . '/../persistence/PersistenceURLMS.php';
require_once $my_dir . '/../model/Staff.php';
require_once $my_dir . '/../model/URLMS.php';
require_once $my_dir . '/../model/Lab.php';
require_once $my_dir . '/../model/Associate.php';
require_once $my_dir . '/../model/Assistant.php';
require_once $my_dir . '/../model/Director.php';
require_once $my_dir . '/../model/Role.php';
require_once $my_dir . '/../model/Equipment.php';
require_once $my_dir . '/../model/Expense.php';
require_once $my_dir . '/../model/FundingAccount.php';
require_once $my_dir . '/../model/Inventory.php';
require_once $my_dir . '/../model/ProgressUpdate.php';
require_once $my_dir . '/../model/Resource.php';
require_once $my_dir . '/../model/Supply.php';




class ManagerController
{
	public function __construct()
	{ //On construction

	    $pm = new PersistenceURLMS();
	    $urlms = $pm->loadDataFromStore(); //Return a new URLMS with a Lab from memory or make a new one
	    $lab = $urlms->getSingleLab();
	    $_SESSION["persistence"] = $pm;
	    $_SESSION["currURLMS"] = $urlms;
	    $_SESSION["currLab"] = $lab;
	    $_SESSION["currStaff"] = $lab->getStaff();
	    $_SESSION["currInventory"] = $lab->getInventory();
	    //$_SESSION["currResources"] =  $_SESSION["currInventory"]->getResources();

	}
	
    /*
     * Inputs are first name, last name, role
     * Inputs received via Submit button from index
     * 
     */
	public function addStaffMember($FirstName, $LastName, $Role) {
		//Instantiate local Variables
	    $Fname = $FirstName;
	    $Lname = $LastName;
	    $role = $Role;
	    
	    //1. First Validate inputs
	    
		if ($Fname == null || strlen($Fname) == 0) {//if empty first name field
		    if ($Lname == null || strlen($Lname) == 0) {//if empty first and last name field
		        throw new Exception("Both First and Last name cannot be empty!");
		    } else{//just empty first name field
			    throw new Exception("First name cannot be empty!");
			}
		} 
		elseif($Lname == null || strlen($Lname) == 0){ //empty last name
		    throw new Exception("Last name cannot be empty!");
		}
		elseif(is_numeric($Fname) || is_numeric($Lname)){ //Numbers in name
		    throw new Exception("Name cannot include numbers");
		}
		else { //Proper names
		   
			// 2. Load all of the data
                //Moved to constructor

			// 3. Add the new Staff Member

			/*array_push($_SESSION["staffLastName"],$LastName);
			if(isset($_SESSION["currLab"])){
			    echo "curr lab is set at least";
			    echo serialize($_SESSION["currLab"]);
			}*/
			    
			$newStaff = new Staff($Fname,$Lname,$_SESSION["currLab"]);//Create new staff object
			//Assign Roles
			if(strcmp($role, "Director")==0){ 
			    if($_SESSION["currLab"]->hasDirector()){
			        throw new Exception("Lab already has a director associated to it");
			    }else{
			        $d = new Director();
			        $newStaff->setRole($d);
			        $_SESSION["currLab"]->setDirector($newStaff);
			    }
			}else if(strcmp($role, "Associate")==0){
			    $a1 = new Associate();
			    $newStaff->setRole($a1);
			}else if(strcmp($role, "Assistant")==0){
			    $a2 = new Assistant();
			    $newStaff->setRole($a2);
			}
			
			$_SESSION["currLab"]->addStaff($newStaff);
			
			// 4. Write all of the data
			$_SESSION["persistence"]->writeDataToStore($_SESSION["currLab"]);
		}
	}
	/*
	 * Takes as input Index number, first name, and last name
	 * Inputs passed via "remove staff" submit button on index page
	 */
	public function removeStaffMember($indexNo,$FirstName, $LastName){
	    $Director = new Director();
	    if($indexNo!=null && !is_numeric($indexNo)){
	        throw new Exception("Index number must be a number");
	    }else if($_SESSION["currLab"]->numberOfStaff()==0){
	        throw new Exception("No staff member exists in this lab");
	    }else if($indexNo!=null && is_numeric($indexNo)){ //If valid index number suplied
	        $staffer = $_SESSION["currLab"]->getStaff_index($indexNo);//Get staff object from current lab
	        if($staffer->getRole()==$Director){	            
	            $_SESSION["currLab"]->setDirector(null);
	        }
	        $staffer->delete();
	    }else{//no index number supplied at all
	        
	        $staffList = $_SESSION["currLab"]->getStaff(); //Get staff list from name
	        $counter = 0;
	        $removeList = array(); //An array of the staff objects to be removed
	        foreach($staffList as $staffer){ //For each staff member     	
	            if ($staffer->getFirstName() == $FirstName && $staffer->getLastName() == $LastName ){
	               	$removeList[$counter] = $staffer; //Add to removeList iff first and last name matches
	               	if($staffer->getRole()==$Director){
	               	    $_SESSION["currLab"]->setDirector(null);
	               	}
	                $counter++;
	            }
	        }
	        if($counter == 1){ //Unique staffmember with name
	            $removeList[0]->delete();
	        } elseif($counter>1){  //I.e. More than 1 person with the same name
	           throw new Exception("Multiple staff members with the same Name, Please input index No in order to remove correct staff member"); 
	        } elseif($counter == 0){ //No staff members with that name
	            throw new Exception("No staff member exists with that name, Either check name input or input index No.");
	        }
	    }
	    $_SESSION["persistence"]->writeDataToStore($_SESSION["currLab"]);
	}
	
	
	/*
	 * Inputs are  name, quantity, type
	 * Inputs received via Submit button from inventoryPage
	 *
	 */
	public function addResource($name, $quantity, $type) {
 	    //Instantiate local Variables
	    $resourcesList = $_SESSION["currLab"]->getInventory()->getResources();
        
	    if($_SESSION["currInventory"]->numberOfResources()>1&&$resourcesList!=null){
	        foreach ($resourcesList as $re ){
	            if($re->getName()==$name) {
	                throw new Exception("The resource already exists in the inventory!");
	            }
	          }     
	    }
	    
	    //1. First Validate inputs
	    if ($name == null || strlen($name) == 0) {//if empty  name field
	        
	        throw new Exception("Name of inventory cannot be empty!");
	        
	    }	   

	    else if($quantity == null || strlen($quantity)==0){ //Numbers in name
	        
	        throw new Exception("Quantity cannot be empty!");
	        
	    }

	    else { 

	        
	        //Check resource type
	        if(strcmp($type, "Supply")==0){
	           //create new resource according to different type
	            $newSupply = new Supply($quantity,$name,$_SESSION["currInventory"]);

	        }else if(strcmp($type, "Equipment")==0){	          
	            $newEquipment = new Equipment($quantity,$name,$_SESSION["currInventory"]);//Create new staff object
	        }else{
	            throw new Exception("Must select a Type");
	        }
	        // 4. Write all of the data
	        $_SESSION["persistence"]->writeDataToStore($_SESSION["currLab"]);
	    }
	}
	

	public function removeResource($index, $name) {
	    $resourcesList = $_SESSION["currLab"]->getInventory()->getResources();
	    $counter = 0;
	    if($index==null && $name==null ) {
	        throw new Exception("Name of the resource cannot be empty!");
	    }
	    else if ($index!=null&&$index>$_SESSION["currLab"]->getInventory()->numberOfResources()){
	        throw new Exception("Index number doesn't exist! ");
	    }
	    else if ($_SESSION["currLab"]->getInventory()->numberOfResources()==0){
	        throw new Exception("No resource exists with that name in the inventory! ");
	    }
	    else if($index!=null&&$index>=0&&$index<=$_SESSION["currLab"]->getInventory()->numberOfResources()){
	        $aResource = $_SESSION["currInventory"]->getResource_index($index);
	        $aResource->delete();        
	    }
	    else {
	     
	        foreach($resourcesList as $aResource){
	            if($aResource->getName()==$name) {
	                $aResource->delete();
	                $counter++;
	            }
	        }
	        if($counter==0) {
	            throw new Exception("No resource exists with that name in the inventory!");
	        }
	    }
	    $_SESSION["persistence"]->writeDataToStore($_SESSION["currLab"]);
	}
	
	public function editResource($name, $quantity) {
	    $resourcesList = $_SESSION["currLab"]->getInventory()->getResources();
	    $counter = 0;
	    if($_SESSION["currInventory"]->numberOfResources()==0&&$resourcesList==null){
	       throw new Exception("No resource exists in the inventory!");  
	    }
	    else {
	        foreach($resourcesList as $aResource){
	            if($aResource->getName()==$name){
	                $aResource->setQuantity($quantity);
	            }
	        }
	    }
	    $_SESSION["persistence"]->writeDataToStore($_SESSION["currLab"]);	    
	}
	
	public function addExpense($description,$amount,$type,$staffLastName) {
	    $fundingAccount = $_SESSION["currLab"]->getFundingAccount();
	    $staffList = $_SESSION["currLab"]->getStaff(); 
	    if ($description == null || strlen($description) == 0) {//if empty  name field
	        
	        throw new Exception("Description cannot be empty!");
	        
	    }
	    
	    else if($amount == null || strlen($amount)==0){ //Numbers in name
	        
	        throw new Exception("Amount cannot be empty!");
	        
	    }
	    
	    else{
	        $date = date("m/d/Y",time());
	        $aExpense=new Expense($description,$amount,$date,$fundingAccount);
	        if($type=="Travel"){
	           $aExpense->setType("TypeTRAVEL") ;
	        }
	        else if($type =="Salary"){
	            $aExpense->setType("TypeSALARY");
	        }
	        else if($type =="Resource"){
	            $aExpense->setType("TypeRESOURCE");
	        }
	        else if($type == "Other"){
	            $aExpense->setType("TypeOTHER");
	        }
	        foreach($staffList as $aStaff){
	            if($aStaff->getLastName()==$staffLastName){
	                $aStaff->addExpense($aExpense);
	            }
	        }

	    }
	    $_SESSION["persistence"]->writeDataToStore($_SESSION["currLab"]);
	}
	
	public function setAllocatedAmount($amount){
	    $fundingAccount = $_SESSION["currLab"]->getFundingAccount();
	    $fundingAccount->setAllocatedAmount($amount);
	    $_SESSION["persistence"]->writeDataToStore($_SESSION["currLab"]);
	}

	public function addProgress($staff,$Message){ //$staff is index of staff array
	    $date = date("m/d/Y", time());//Format date into m/d/Y based of UNIX time
	    $progStaff = $_SESSION["currLab"]->getStaff_index($staff);
	    $first = $progStaff->getFirstName();
	    $last = $progStaff->getLastName();
	    $role = $progStaff->getRole();
	    $role = get_class($role);
	    $progStaffName = $first." ".$last.": ";
	    $message = $staff.".".$progStaffName.$role.";\n".$date."\n \n".$Message; //To Be placed inside of the view page so that the view shows staff member as well as message
	    new ProgressUpdate($message,$date,$_SESSION["currLab"]);
	    $_SESSION["persistence"]->writeDataToStore($_SESSION["currLab"]);//Save to persistence
	}
	

}
 
?>
